### ISSUE TITLE
Wrong outputs from IRFFT on GPU

### ISSUE THREAD

**Issue by kohr-h** [google/jax#2874]
It's kinda hard to describe what's going on. I've tried to narrow it down as much as possible, but at this point I can't think of many more ways.

Also, I've tried to come up with a reasonably simple example, but it's probably possible to simplify it some more.

**Problem Description:**

Imports
```py
from functools import partial

import matplotlib
import matplotlib.pyplot as plt
import numpy as np
import jax
from jax import jit
from jax import numpy as jnp

matplotlib.rc("image", cmap="gray")
```

Let's define some functions.
```py
# Probably none of the specifics of these functions matter,
# since JIT will turn the return value to a constant, as seen
# by checking the jaxpr

@partial(jit, static_argnums=(0, 1))
def create_gaussian(shape, sigma):
    xs = jnp.arange(shape[0], dtype="float32") - shape[0] // 2
    ys = jnp.arange(shape[1], dtype="float32") - shape[1] // 2
    xx = xs[:, None]
    yy = ys[None, :]
    return jnp.exp((xx ** 2) / (-2 * sigma[0] ** 2) + (yy ** 2) / (-2 * sigma[1] ** 2))


@partial(jit, static_argnums=(0, 1))
def create_gaussian_rfft(shape, sigma):
    return jnp.fft.rfftn(jnp.fft.ifftshift(create_gaussian(shape, sigma)))
```
The result of the *first run* of the following code:
```py
gaussian = jnp.fft.fftshift(jnp.fft.irfftn(create_gaussian_rfft((2048, 2048), (2.0, 5.0))))
gaussian_np = np.asarray(gaussian)
n, m = gaussian.shape

fig, ax = plt.subplots(figsize=(10, 10))
ax.imshow(gaussian_np)
```
looks fine:
![good](https://user-images.githubusercontent.com/5030250/80532835-04b40980-899d-11ea-921f-6bb7a6ea3136.png)

But when running the *exact same code again* (not including the function definitions) here's what I get:
![bad](https://user-images.githubusercontent.com/5030250/80532943-28774f80-899d-11ea-8efd-fd1002308c08.png)

After that the outcome will alternate between those two versions.

I've tried to eyeball the width of the Gaussian in axis 0, it's about 160. Funny thing is that if I put 160 instead of 2, it seems to alternate between the same versions, only starting with the wide one.

**Things that fix it:**

- This behavior occurs on a Volta GPU (Titan V) with CUDA 10.2 and driver version 440.x. I mention that because I tried to reproduce the issue on another machine (Pascal GPU, Titan Xp) with an older driver version 418.x, without success.
  (I would have tested the Volta system with older driver versions and/or older CUDA, but I'm using the Ubuntu repo, and it doesn't like downgrading the driver while keeping CUDA somewhat recent.)
- Likewise, the issue does not occur when I use an array or a list instead of a tuple for the `sigma` parameter. But I guess that's because it triggers recompilation.
- Removing the FFT from the picture and just generating the array also eliminates the issue.
- The non-jitted versions also don't exhibit the problem.
- And finally, making the `sigma` argument of `create_gaussian_rfft` dynamic by using `static_argnums=(0,)` also makes the problem go away.

**Comment by hawkinsp** [comment]
I think this is an XLA/GPU bug caused by irfft mutating its inputs. Here's a simpler repro:
```
import jax.numpy as jnp
import numpy as np

x = np.random.randn(2048, 1025) + np.random.randn(2048, 1025) * 1j
y = jnp.asarray(x)
print(y)
jnp.fft.irfftn(y)
print(y + 0.)  # Note the addition to force a new copy rather than caching.
jnp.fft.irfftn(y)
print(y + 0.)  # Note the addition to force a new copy rather than caching.
```
In theory this should print the same value three times. In practice it does not.

I filed an (internal) issue with the XLA folks.

**Comment by hawkinsp** [comment]
(A workaround would be to use CPU.)

**Comment by kohr-h** [comment]
Thanks @hawkinsp for your quick answer. I can confirm that your simpler repro behaves the same as my example (results differ on Volta, but not on Pascal).

Any idea why this depends on GPU architecture? Do the FFT algos differ between them?

**Comment by hawkinsp** [comment]
Our current hypothesis is this is a bug in CuFFT, which would explain why it depends on the GPU architecture.

**Comment by kohr-h** [comment]
Just found this bit in the ["Data Layout" section of the cuFFT documentation](https://docs.nvidia.com/cuda/cufft/index.html#data-layout):

    Note that in-place complex-to-real FFTs may overwrite arbitrary imaginary input point
    values when non-unit input and output strides are chosen. Out-of-place complex-to-real
    FFT will overwrite input buffer if custom strides are set by the user.

**Comment by kohr-h** [comment]
Possibly the "may" is the architecture-dependent part.

**Comment by hawkinsp** [comment]
Nice find! That's a surprising definition of "out-of-place", to say the least.

**Comment by kohr-h** [comment]
Indeed. This should be a big red warning in the docs.

To be fair, FFTW has [the same behavior](http://www.fftw.org/fftw3_doc/Planner-Flags.html#Planner-Flags) (see under `FFTW_PRESERVE_INPUT`) and doesn't necessarily do a better job at warning about it.

**Comment by kohr-h** [comment]
Is there a solution in sight on this issue? For my application it blocks me from using RFFT on Volta.

**Comment by hawkinsp** [comment]
I've asked the XLA folks to work around the NVIDIA IRFFT bug by cloning the input for all IRFFTs. Hopefully that can happen relatively quickly.

(If any NVIDIA folks are reading this, this is partners bug # 2959622 )

Interestingly PyTorch also seems to have experienced the same bug: https://github.com/pytorch/pytorch/issues/34551

**Comment by hawkinsp** [comment]
I believe this is fixed with jaxlib 0.1.51. The XLA folks added a workaround which is to copy the input buffer for IRFFTs. Hopefully NVidia will fix this problem and we can avoid the copy in the future, but this fixes the user-visible problem at least.

Please reopen if you are still having problems!
